File: Java/Hard/1622 Fancy Sequence.java
class Fancy {
    private List<Long> nums;
    private long multiplier = 1;
    private long increment = 0;
    private static final int MOD = 1_000_000_007;

    public Fancy() {
        nums = new ArrayList<>();
    }
    
    public void append(int val) {
        // Reverse the current transformation: (val - increment) / multiplier
        long transformed = (val - increment + MOD) % MOD;
        transformed = (transformed * modInverse(multiplier, MOD)) % MOD;
        nums.add(transformed);
    }
    
    public void addAll(int inc) {
        increment = (increment + inc) % MOD;
    }
    
    public void multAll(int m) {
        multiplier = (multiplier * m) % MOD;
        increment = (increment * m) % MOD;
    }
    
    public int getIndex(int idx) {
        if (idx >= nums.size()) return -1;
        // Apply the current transformation: (transformed_val * multiplier) + increment
        long res = (nums.get(idx) * multiplier) % MOD;
        res = (res + increment) % MOD;
        return (int) res;
    }

    private long modInverse(long n, int mod) {
        return power(n, mod - 2, mod);
    }

    private long power(long x, int y, int m) {
        long res = 1;
        x %= m;
        while (y > 0) {
            if (y % 2 == 1) res = (res * x) % m;
            y = y >> 1;
            x = (x * x) % m;
        }
        return res;
    }
}

File: C/Hard/1622 Fancy Sequence.c
#include <stdlib.h>
#include <stdio.h>

#define MOD 1000000007

typedef struct {
    long long *nums;
    int size;
    int capacity;
    long long multiplier;
    long long increment;
} Fancy;

// Helper: Power function for Modular Inverse (Fermat's Little Theorem)
long long power(long long base, long long exp) {
    long long res = 1;
    base %= MOD;
    while (exp > 0) {
        if (exp % 2 == 1) res = (res * base) % MOD;
        base = (base * base) % MOD;
        exp /= 2;
    }
    return res;
}

long long modInverse(long long n) {
    return power(n, MOD - 2);
}

Fancy* fancyCreate() {
    Fancy* obj = (Fancy*)malloc(sizeof(Fancy));
    obj->capacity = 100;
    obj->nums = (long long*)malloc(sizeof(long long) * obj->capacity);
    obj->size = 0;
    obj->multiplier = 1;
    obj->increment = 0;
    return obj;
}

void fancyAppend(Fancy* obj, int val) {
    if (obj->size == obj->capacity) {
        obj->capacity *= 2;
        obj->nums = (long long*)realloc(obj->nums, sizeof(long long) * obj->capacity);
    }
    
    // Reverse the transformation: x = (val - inc) * inv(mult)
    long long transformed = (val - obj->increment + MOD) % MOD;
    transformed = (transformed * modInverse(obj->multiplier)) % MOD;
    
    obj->nums[obj->size++] = transformed;
}

void fancyAddAll(Fancy* obj, int inc) {
    obj->increment = (obj->increment + inc) % MOD;
}

void fancyMultAll(Fancy* obj, int m) {
    obj->multiplier = (obj->multiplier * m) % MOD;
    obj->increment = (obj->increment * m) % MOD;
}

int fancyGetIndex(Fancy* obj, int idx) {
    if (idx >= obj->size) return -1;
    
    // Apply the current transformation: (stored_val * mult + inc)
    long long res = (obj->nums[idx] * obj->multiplier) % MOD;
    res = (res + obj->increment) % MOD;
    return (int)res;
}

void fancyFree(Fancy* obj) {
    free(obj->nums);
    free(obj);
}
